### PR TITLE
Add llama model to benchmark

### PR DESCRIPTION
**Pull request by xadupre**

### CODE REVIEW

File: onnxscript/tools/transformers_models/llama.py [hunk 1]
```diff
@@ -92,3 +92,75 @@ def generate_example_inputs(batch: int, seq: int, vocab_size: int):
...
+    num_hidden_layers: int = 1,
+    implementation: str = "eager",
+    dynamic_shapes: bool = False,
+    with_mask: bool = True,
+) -> tuple[Any, list[tuple[torch.Tensor, ...]], dict]:
+    """
+    Returns a model Phi to test or benchmark.
+
+    Args:
+        warmup: number of inputs to generate
```
Comment (justinchuby) [inline onnxscript/tools/transformers_models/llama.py:110]: no big deal but for future reference: Capitalize the start of the phase and end with a period. See https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html (note we don't need to include type information as the example did because we have type info in the function signature already).

`warmup: Number of inputs to generate.`

File: onnxscript/tools/transformers_models/llama.py [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+        return (input_ids,)
+
+    example_args_collection = []
+    for b, s in input_dims:
+        example_args_collection.append(generate_example_inputs(b, s, vocab_size))
+
+    return LlamaModelWrapper(config), example_args_collection, dynamic_shapes
+
+
+def get_llama_model_config(
```
Comment (justinchuby) [inline onnxscript/tools/transformers_models/llama.py:97]:
```suggestion
def get_llama_model_from_config(
```

Otherwise it looks like it's producing a config.

File: onnxscript/tools/transformers_models/llama.py [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+            num_hidden_layers=num_hidden_layers,
+            vocab_size=32000,
+            intermediate_size=11008,
+            max_position_embeddings=2048,
+            num_attention_heads=32,
+            _attn_implementation=implementation,
+            with_mask=with_mask,
+        )
+    else:
+        raise AssertionError(f"Unexpected configuration {config!r}.")
```
Comment (justinchuby) [inline onnxscript/tools/transformers_models/llama.py:164]:
```suggestion
        raise ValueError(f"Unexpected configuration {config!r}.")
```

File: onnxscript/tools/transformers_models/llama.py [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+    for b, s in input_dims:
+        example_args_collection.append(generate_example_inputs(b, s, vocab_size))
+
+    return LlamaModelWrapper(config), example_args_collection, dynamic_shapes
+
+
+def get_llama_model_config(
+    warmup: int = 5,
+    repeat: int = 10,
+    config: str = "small",
```
Comment (justinchuby) [inline onnxscript/tools/transformers_models/llama.py:100]:
```suggestion
    config: Literal["small", "medium", "large"] = "small",
```

File: onnxscript/tools/transformers_models/llama.py [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+    num_hidden_layers: int = 1,
+    implementation: str = "eager",
```
Comment (justinchuby) [inline onnxscript/tools/transformers_models/llama.py:102]:
```suggestion
    implementation: Literal["eager", "sdpa"] = "eager",
```

File: onnxscript/tools/transformers_models/llama.py (hunk 1 above)
Comment (xadupre) [reply onnxscript/tools/transformers_models/llama.py]: done

File: onnxscript/tools/transformers_models/llama.py (hunk 2 above)
Comment (xadupre) [reply onnxscript/tools/transformers_models/llama.py]: done

File: onnxscript/tools/transformers_models/llama.py (hunk 4 above)
Comment (xadupre) [reply onnxscript/tools/transformers_models/llama.py:100]: ruff returns: F821 Undefined name `large`, the list may evolves so I prefer to stick with a string

File: onnxscript/tools/transformers_models/llama.py (hunk 3 above)
Comment (xadupre) [reply onnxscript/tools/transformers_models/llama.py]: done

File: onnxscript/tools/transformers_models/llama_test.py [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+import torch
+
+import onnxscript.tools.training_helper
+import onnxscript.tools.transformers_models
+import onnxscript.tools.transformers_models.llama
+
+HAS_TRANSFORMERS = onnxscript.tools.transformers_models.has_transformers()
+
+
+class TestExportPhi(unittest.TestCase):
```
Comment (gramalingam) [inline onnxscript/tools/transformers_models/llama_test.py:20]: Should this be `TestExportLlama` ?

**Review by gramalingam — APPROVED** [review]
Just one minor renaming nit

File: onnxscript/tools/transformers_models/llama_test.py (hunk 6 above)
Comment (xadupre) [reply onnxscript/tools/transformers_models/llama_test.py:20]: true

File: onnxscript/tools/transformers_models/__init__.py [hunk 7]
```diff
@@ -22,6 +27,38 @@ def has_transformers():
         return False
 
 
+def has_torch(version: str = ""):
+    """Tells if pytorch is installed and recent enough."""
+    import packaging.version as pv
```
Comment (justinchuby) [inline onnxscript/tools/transformers_models/__init__.py:32]: nit: Avoid unconventional abbreviations. I would just type the whole thing.

File: onnxscript/tools/transformers_models/__init__.py [hunk 8]
```diff
@@ -22,6 +27,38 @@ def has_transformers():
         return False
 
 
+def has_torch(version: str = ""):
```
Comment (justinchuby) [inline onnxscript/tools/transformers_models/__init__.py:30]: Just wondering: do these two functions belong in this module? I assume we have more flexible rules for tools/ but we need to communicate that these are not going to be stable apis.

Comment (xadupre) [reply onnxscript/tools/transformers_models/__init__.py:30]: I can put them in onnxscript/testing. transformers and pytorch changes a lot. I expect some tests to only work with recent versions of both.

Comment (justinchuby) [reply onnxscript/tools/transformers_models/__init__.py:30]: Not suggesting onnxscript/testing because that's a set of more stable API we provide. I would consider `onnxscript._internal.testing`

File: onnxscript/tools/transformers_models/__init__.py (hunk 7 above)
Comment (justinchuby) [reply onnxscript/tools/transformers_models/__init__.py:32]: I would also move the import out. We should have inline imports like this only in rare cases (e.g. torch/transformers etc. in command line utils)

Comment (xadupre) [reply onnxscript/tools/transformers_models/__init__.py]: removed

File: onnxscript/tools/benchmark/export_model_test.py [hunk 9]
```diff
@@ -6,7 +6,11 @@
 import unittest
 
 import onnxscript.tools.benchmark.export_model
-from onnxscript.tools.transformers_models import has_transformers
+from onnxscript._internal.version_utils import (
+    has_transformers,
+    is_onnxruntime_training,
+    torch_older_than,
+)
```
Comment (justinchuby) [inline onnxscript/tools/benchmark/export_model_test.py:13]:
```suggestion
from onnxscript._internal import version_utils
```

nit

File: onnxscript/tools/transformers_models/llama_test.py [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+import unittest
+
+import numpy as np
+import onnxruntime
+import torch
+
+import onnxscript.tools.training_helper
+import onnxscript.tools.transformers_models
+import onnxscript.tools.transformers_models.llama
+from onnxscript._internal.version_utils import has_transformers
```
Comment (justinchuby) [inline onnxscript/tools/transformers_models/llama_test.py:16]:
```suggestion
from onnxscript._internal import version_utils
```

File: onnxscript/tools/transformers_models/phi_test.py [hunk 11]
```diff
@@ -16,32 +15,18 @@
...
-    if optimize:
-        model_proto = onnxscript.optimizer.optimize(
-            model_proto,
-            num_iterations=2,
-            onnx_shape_inference=True,
-        )
-        model_proto = onnxscript.rewriter.rewrite(model_proto)
-        model_proto = onnx.inliner.inline_local_functions(model_proto)
-    return model_proto
+from onnxscript._internal.version_utils import has_transformers, torch_older_than
```
Comment (justinchuby) [inline onnxscript/tools/transformers_models/phi_test.py:18]:
```suggestion
from onnxscript._internal import version_utils
```

**Review by justinchuby — APPROVED** [review]
LGTM with minor comments

File: noxfile.py [hunk 12]
```diff
@@ -73,10 +77,13 @@ def test_torch_nightly(session):
     session.install(
         *COMMON_TEST_DEPENDENCIES,
         ONNX_RUNTIME,
+        TRANSFORMERS,
     )
     session.install("-r", "requirements/ci/requirements-onnx-weekly.txt")
     session.install("-r", "requirements/ci/requirements-pytorch-nightly.txt")
     session.install(".", "--no-deps")
+    if sys.platform == "linux":
+        session.install("", "onnxruntime-training")
```
Comment (justinchuby) [inline noxfile.py:86]: I would call this out in the docstring under `test_torch_nightly` to make it very explicit what we are doing.

File: onnxscript/tools/transformers_models/phi_test.py (hunk 11 above)
Comment (xadupre) [reply onnxscript/tools/transformers_models/phi_test.py:18]: The lines become longer and the linter splits the lines over multiple lines and it is difficult to read. That's why I did it that way.

File: onnxscript/tools/benchmark/export_model_test.py (hunk 9 above)
Comment (xadupre) [reply onnxscript/tools/benchmark/export_model_test.py:13]: The lines become longer and the linter split them over multiple line. It is easier to read that way.

File: noxfile.py (hunk 12 above)
Comment (xadupre) [reply noxfile.py]: done

File: tests/function_libs/torch_lib/ops_test_data.py [hunk 13]
```diff
@@ -842,7 +842,10 @@ def _where_input_wrangler(
         matcher=lambda sample: ("dtype" in sample.kwargs),
         reason="this Aten overload only support dtype not in kwargs",
     ),
-    TorchLibOpInfo("gather", core_ops.aten_gather),
+    TorchLibOpInfo("gather", core_ops.aten_gather).skip(
+        enabled_if=not version_utils.torch_older_than("2.4"),
```
Comment (justinchuby) [inline tests/function_libs/torch_lib/ops_test_data.py:846]: This is a failure we need to fix @fatcat-z @xiaowuhu
